---
title: "Week 2 Assignment"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error = TRUE, warning = TRUE)

library(ottr)
library(testthat)
```

---

## Overview

---

This assignment covers the content in week 2 lectures and tutorials. 

#### Mark breakdown
- Question 1 (5 marks)
- Question 2 (10 marks)
- Question 3 (8 marks)

---

## Background

---

Data in this assignment are based on information from:

The ENCODE Project Consortium., Moore, J.E., Purcaro, M.J. et al. Expanded encyclopaedias of DNA elements in the human and mouse genomes. Nature 583, 699â€“710 (2020).

Promoter and enhancer regions regulate the transcription of genes.

- A *promoter region* is a sequence just upstream of the transcription start site (TSS) of a gene that enables transcription initiation. 

- An *enhancer region* is a sequence that influences the rate of transcription of a particular gene. It does not have a specific location with respect to the gene it regulates, although they are cis-acting (on the same chromosome). A *proximal enhancer* is an enhancer that is close the TSS.

Promoters and enhancers have the following characteristics:

|                   | Promoter region | Enhancer region      |
| ----------------- | --------------- | -------------------- |
| Distance from TSS | < 200 bp        | no specific distance |
| DHS               | TRUE            | TRUE                 |
| H3K4me3 level     | high            | low                  |
| H3K27ac level     | low             | high                 |


- *DHS*: active promoter and enhancer regions contain DNAse I hypersensitivity sites (DHSs). In these data, regions that contain DHSs will have a "DHS" value of TRUE, indicating that the regions could be a promoter or an enhancer.

- *H3K4me3*: a histone mark that is present at a HIGH level in promoter regions and a LOW level in enhancer regions

- *H3K27ac*: a histone mark that is present at a LOW level in promoter regions and a HIGH level in enhancer regions

We will start by examining three regions in the human genome: A, B, and C. Characteristics for these three regions will be assigned to the following variables (X represents A, B, or C):

- `rX_dist_TSS`     distance from closest TSS (numeric)
- `rX_DHS`          is it a DNAse I hypersensitivity site (logical)
- `rX_H3K4me3`      level of H3K4me3: high or low (character)
- `rX_H3K27ac`      level of H3K27ac: high or low (character)

```{r}
# values for region A
rA_dist_TSS <- 158
rA_DHS <- TRUE
rA_H3K4me3 <- "high"
rA_H3K27ac <- "low"

# values for region B
rB_dist_TSS <- 192
rB_DHS <- FALSE
rB_H3K4me3 <- "low"
rB_H3K27ac <- "low"

# values for region C
rC_dist_TSS <- 4729
rC_DHS <- TRUE
rC_H3K4me3 <- "low"
rC_H3K27ac <- "high"
```

---

## QUESTIONS 

---

### Question 1

#### a. 
(1 mark)
Region A is closest to the TSS of the gene "MBNL1". Convert the variable `rA_gene_lowercase` defined below to uppercase letters using one of the functions we learned this week. Assign the value to a variable named "q.1.a".

```{r Assignment.W2.1.a}
rA_gene_lowercase <- "mbnl1"
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.1.a.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### b.
(1 mark)
Using one of the functions we learned this week, get the number of letters in the value saved to variable `rA_H3K27ac` (i.e., not in the variable name "rA_H3K27ac" itself). Assign the result to a variable named "q.1.b".

```{r Assignment.W2.1.b}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.1.b.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### c. 
(1 mark)
The mean distance of regions A, B, and C from the TSS that is closest to each of them is calculated below and assigned to the variable `mean_dist_TSS`. Using one of the functions we learned this week, round `mean_dist_TSS` to two digits after the decimal place. Assign the return value to a variable named "q.1.c".

```{r Assignment.W2.1.c}
mean_dist_TSS <- (rA_dist_TSS + rB_dist_TSS + rC_dist_TSS)/3
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.1.c.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### d.
(2 marks)
By nesting two functions that we learned this week, extract the 2nd and 3rd characters of the value saved to variable `rC_H3K4me3` and output a logical value that indicates whether those 2 letters are also found in the value saved to variable `rA_H3K27ac` (i.e., do not compare the variable names). Assign the return value to a variable named "q.1.d".

```{r Assignment.W2.1.d}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.1.d.R")
```
**IGNORE the above block: test results will not appear in assignments** 

---

### Question 2

#### a.
(1 mark)
Write a logical expression that answers the following question: Is region B close enough to a TSS for region B to be a promoter? 
Remember: promoters must be within 200 bp of the TSS.
Assign the return value of the logical expression to a variable named "q.2.a".

```{r Assignment.W2.2.a}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.2.a.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### b.
(1 mark)
Write a logical expression that answers the following question: Is the level of H3K27ac the same in region B as it is region C? 
Assign the return value of the logical expression to a variable named "q.2.b".

```{r Assignment.W2.2.b}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.2.b.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### c.
(1 mark)
Write a logical expression that answers the following question: Is the distance of region C from the closest TSS greater than or equal to the distance from the closest TSS for region B? 
Assign the return value of the logical expression to a variable named "q.2.c".

```{r Assignment.W2.2.c}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.2.c.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### d.
(1 mark)
Write a logical expression that answers the following question: Is region A close enough to a TSS for region A to be a promoter and is region B close enough to a TSS for region B to be a promoter?
Assign the return value of the logical expression to a variable named "q.2.d".

```{r Assignment.W2.2.d}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.2.d.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### e. 
(1 mark)
Write a logical expression that answers the following question: Do region B or region C have a high level of H3K27ac?
Assign the return value of the logical expression to a variable named "q.2.e".

```{r Assignment.W2.2.e}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.2.e.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### f. 
(1 mark)
Write a logical expression that answers the following question: Does region C NOT have a high level of H3K4me3 AND does it NOT have low level of H3K4me3?
Assign the return value of the logical expression to a variable named "q.2.f".

```{r Assignment.W2.2.f}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.2.f.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### g. 
(2 marks)
Write a logical expression that answers the following question: Do any of the regions (A, B, or C) NOT contain DNAse hypersentivity sites?
Assign the return value of the logical expression to a variable named "q.2.g".

```{r Assignment.W2.2.g}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.2.g.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### h. 
(2 marks)
Write a logical expression that answers the following question: Does region A or region B contain DNAse hypersentivity sites AND does region A or region B have a high level of H3K27ac?
Assign the return value of the logical expression to a variable named "q.2.h".

```{r Assignment.W2.2.h}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.2.h.R")
```
**IGNORE the above block: test results will not appear in assignments** 

---

### Question 3

#### a. 
(1 mark) 
Write an if statement that checks if region C contains DNAse hypersensitivity sites and if it does, prints "Region is likely a promoter or an enhancer". Do not alter the line containing the code `q.3.a <-` or you will not receive a mark for this question.

```{r Assignment.W2.3.a}
q.3.a <- # Do not modify this line
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.3.a.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### b. 
(1 mark)
Write an if-else statement that checks if region B is within 200 base pairs of a TSS. If it is within 200 base pairs print "Region is potentially a promoter", otherwise print "Region cannot be a promoter". Do not alter the line containing the code `q.3.b <-` or you will not receive a mark for this question.

```{r Assignment.W2.3.b}
q.3.b <- # Do not modify this line
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.3.b.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### c.
(3 marks)
Use one or more conditional statements to write code that checks the values of the 4 arguments defined below (`rX_DHS`, `rX_dist_TSS`, `rX_H3K4me3`, `rX_H3K27ac`) to determine if the region is likely a promoter, likely an enhancer, or neither. 

- If the variables match the criteria for a promoter, the value "promoter" should be assigned to a variable named "region_type". 
- If the variables match the criteria for an enhancer, the value "enhancer" should be assigned to a variable named "region_type". 
- If the variables DO NOT match the criteria for a promoter and DO NOT match the criteria for an enhancer, the value "unknown" should be assigned to a variable named "region_type". 

Your code will be checked with different values of `rX_DHS`, `rX_dist_TSS`, `rX_H3K4me3`, and `rX_H3K27ac`, so make sure it will work for ALL possible situations.

Here is a reminder of the characteristics of promoter and enhancer regions:

|                   | Promoter region | Enhancer region      |
| ----------------- | --------------- | -------------------- |
| Distance from TSS | < 200 bp        | no specific distance |
| DHS               | TRUE            | TRUE                 |
| H3K4me3 level     | high            | low                  |
| H3K27ac level     | low             | high                 |

```{r Assignment.W2.3.c}
rX_DHS <- TRUE
rX_dist_TSS <- 398
rX_H3K4me3 <- "low"
rX_H3K27ac <- "high"
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.3.c.R")
```
**IGNORE the above block: test results will not appear in assignments** 

#### d.
(3 marks)
Use one or more conditional statements to write code that checks if region C is an enhancer and if so, checks if it is a *distal enhancer* (> 2000 bases from the TSS) or a *proximal enhancer* (<= 2000 bases from the TSS). 

- If the region is an enhancer, the value `TRUE` should be assigned to a variable named "is.enhancer".
- If the region is NOT an enhancer, the value `FALSE` should be assigned to a variable named "is.enhancer".
- If the region is an enhancer and the enhancer is distal, the value "distal" should be assigned to a variable named "enhancer.type".
- If the region is an enhancer and the enhancer is proximal, the value "proximal" should be assigned to a variable named "enhancer.type".

```{r Assignment.W2.3.d}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Assignment.W2.3.d.R")
```
**IGNORE the above block: test results will not appear in assignments** 

---